content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+  }
+
+  jfrEvents.awaitEvents();
+
+  long count = jfrEvents.filter(this::isDatabaseIoEvent).count(); // <2>
+  assertThat(count / ITERATIONS).isEqualTo(4).describedAs("write + read per statement, write + read per commit");
+
+  long bytesReadOrWritten = jfrEvents.filter(this::isDatabaseIoEvent)
+    .mapToLong(this::getBytesReadOrWritten)
+    .sum();
+
+  assertThat(bytesReadOrWritten / ITERATIONS).isLessThan(250); // <3>
+}
+
+private boolean isDatabaseIoEvent(RecordedEvent re) { // <4>
+  return ((re.getEventType().getName().equals("jdk.SocketRead") ||
+      re.getEventType().getName().equals("jdk.SocketWrite")) &&
+      re.getInt("port") == databasePort);
+}
+
+private long getBytesReadOrWritten(RecordedEvent re) { // <5>
+  return re.getEventType().getName().equals("jdk.SocketRead") ? re.getLong("bytesRead") : re.getLong("bytesWritten");
+}
+----
+<1> Enable the `jdk.SocketRead` and `jdk.SocketWrite` events; by default, those contain the stacktrace of the read or write event, so that needs to be enabled explicitly
+<2> There should be four events per invocation of the API method
+<3> Less than 250 bytes I/O are expected per invocation
+<4> Only read and write events on the database port are relevant for this test, but e.g. not I/O on the web port of the application
+<5> Retrieve the value of the event's `bytesRead` or `bytesWritten` field, depending on the event type
+
+Now let's again assume that after some time the test begins to fail.
+This time it's the assertion on the number of executed reads and writes:
+
+[source]
+----
+AssertionFailedError: 
+Expecting:
+ <18L>
+to be equal to:
+ <4L>
+but was not.
+----
+
+Also the number of bytes read and written has substantially increased:
+
+[source]
+----
+java.lang.AssertionError: 
+Expecting:
+ <1117L>
+to be less than:
+ <250L>
+----
+
+That's definitely something to look into, as increased database I/O implies increased request latency.
+So let's open the recording of the failed test in Flight Recorder and take a look at the socket read and write events.
+Thanks to enabling stacktraces for the two JFR event types we can quite quickly identify the events asssociated to an invocation of the `GET /todo/{id}` API:
+
+image::continuous_perf_testing_socket_regression.png[Socket read and write events after the performance regression]
+
+At this point, some familiarity with the application in question will come in handy to identify suspicous events.
+But even without that, we could compare previous recordings of successful test runs with the recording from the failing one to see where differences are.
+In the case at hand, the `BlobInputStream` and Hibernate's `BlobTypeDescriptor` in the call stack seem pretty unexpected,
+as our `User` entity didn't have any `BLOB` attribute before.
+
+In reality, comparing with the latest version and a look into the git history of that class could confirm that there's a new attribute storing an image
+(surely not a best practice to do so ;):
```

Comment:
```suggestion

```

I don't think it's that bad. Ideally, make it a lazy blob...

Reply:
Yes, mentioning that as an option later on.